torch/fx/graph_module.py

```diff
@@ -277,6 +266,165 @@ def _gen_model_repr(module_name: str, module: torch.nn.Module) -> Optional[str]:
+        """
+        atoms: List[str] = target.split(".")
+
+        for item in atoms:
+
+            if not hasattr(self, item):
+                return False
+
+            self = getattr(self, item)
+
+            if not isinstance(self, torch.nn.Module):
+                return False
+
+        return True
+
+    def insert_submodule(self, target: str, m: torch.nn.Module) -> bool:
+        """
+        Adds the given submodule to ``self``.
+
+        This installs empty Modules where none exist yet if they are 
+        subpaths of ``target``.
+
+        Args:
+            target: The fully-qualified string name of the new submodule
+                (See example in ``has_submodule`` for how to specify a
+                fully-qualified string.)
+            m: The submodule itself; the actual object we want to
+                install in the current GraphModule
+
+        Return:
+            bool: Whether or not the submodule could be inserted. For
+                this method to return True, each object in the chain
+                denoted by ``target`` must either a) not exist yet,
+                or b) reference an ``nn.Module`` (not a parameter or
+                other attribute)
+
+        """
+        *prefix, field = target.split('.')
+        mod: torch.nn.Module = self
+
+        for item in prefix:
+
+            submod = getattr(self, item, None)
+
+            if submod is None:
+                submod = torch.nn.Module()
+                setattr(mod, item, submod)
+
+            if not isinstance(submod, torch.nn.Module):
+                return False
+
+            mod = submod
+
+        setattr(self, field, m)
+        return True
+
+    def delete_submodule(self, target: str) -> bool:
+        """
+        Deletes the given submodule from ``self``.
+
+        The module will not be deleted if ``target`` is not a valid
+        target.
+
+        Args:
+            target: The fully-qualified string name of the new submodule
+                (See example in ``has_submodule`` for how to denote a
+                fully-qualified string.)
+
+        Returns:
+            bool: Whether or not the target string referenced a
+                submodule we want to delete. A return value of ``False``
+                means that the ``target`` was not a valid reference to
+                a submodule.
+        """
+        def find_and_delete(o: Any, target: str) -> bool:
+            atoms = target.split(".", 1)
+            prefix, path = atoms[0], atoms[1] if len(atoms) > 1 else None
+
+            if not hasattr(o, prefix):
+                return False
+
+            # If the submodule we're looking for is on this layer
+            if not path:
+                if not isinstance(getattr(o, prefix), torch.nn.Module):
+                    return False
+
+                delattr(o, prefix)
+                return True
+
+            return find_and_delete(getattr(o, prefix), path)
+
+        return find_and_delete(self, target)
```

Comment:
This function doesn't need to be recursive anymore, right? Basically just iteratively traverse until the 2nd to last atom then delete the last atom from there